runtime/compiler/optimizer/TreeLowering.cpp

```diff
@@ -467,24 +467,51 @@ AcmpTransformer::lower(TR::Node * const node, TR::TreeTop * const tt)
       regDepForStoreNode->setAndIncChild(0, storeNode->getFirstChild());
       }
 
-   // Using a similar strategy as above, insert check for lhs == NULL.
    auto* const nullConst = TR::Node::aconst(0);
-   auto* const checkLhsNull = TR::Node::createif(TR::ifacmpeq, anchoredCallArg1TT->getNode()->getFirstChild(), nullConst, targetBlock->getEntry());
-   exitGlRegDeps = copyBranchGlRegDepsAndSubstitute(checkLhsNull, exitGlRegDeps, regDepForStoreNode);
-   tt->insertBefore(TR::TreeTop::create(comp, checkLhsNull));
-   callBlock = splitForFastpath(callBlock, tt, targetBlock);
-   if (trace())
-      traceMsg(comp, "Added check node n%un; call node is now in block_%d\n", checkLhsNull->getGlobalIndex(), callBlock->getNumber());
+   bool hasSubstitutedRegDepForSotreNode = false;
+   // Using a similar strategy as above, insert check for lhs == NULL.
+   if (!anchoredCallArg1TT->getNode()->getFirstChild()->isNonNull())
+      {
+      auto* const checkLhsNull = TR::Node::createif(TR::ifacmpeq, anchoredCallArg1TT->getNode()->getFirstChild(), nullConst, targetBlock->getEntry());
+      exitGlRegDeps = copyBranchGlRegDepsAndSubstitute(checkLhsNull, exitGlRegDeps, regDepForStoreNode);
+      hasSubstitutedRegDepForSotreNode = true;
```

Comment:
You might consider eliminating the `hasSubstitutedRegDepForSotreNode` variable, and instead clear the value of `regDepForStoreNode` with:

```
regDepForStoreNode = NULL;
```

Then, below, instead of checking `if (hasSubstitutedRegDepForSotreNode)`, you could do:

```
         exitGlRegDeps = copyBranchGlRegDepsAndSubstitute(checkRhsNull, exitGlRegDeps, regDepForStoreNode);
         regDepForStoreNode = NULL;
```

That takes care of ensuring the `regDepForStoreNode` is only copied once.  I believe it's OK to overwrite `exitGlRegDeps` once the third argument is `NULL`, because register dependencies will be a true copy if there is no substitution to perform.

That would slightly simplify the logic in this method, though the disadvantage is that it might not be obvious that the `exitGlRegDeps` doesn't really change anymore once `regDepForStoreNode` gets set to `NULL`.

Anyway, I'll leave it to you to decide whether to try that suggestion.

Reply:
That's a good idea. I removed `hasSubstitutedRegDepForSotreNode` and it's under test.

Reply:
The change is made in 601ffec